Rapid Piping Device

gusxyz opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
Brings the Rapid Piping Device to space station 14 at MVP state.
Please submit criticism and suggestions.
Highly recommend alongside this changing any steel in Atmos on all maps to an extra RCD charge.

- [ ] Add bent radiators
- [ ] Fix the CONFLICTS :)
- [ ] Requires #34781.
- [ ] Refactor RPD specific code based off review(s).

## Why / Balance
Building pipes in atmos is pain.

## Technical details
Just an RCD with a different set of objects to build.
Changed all references in code to RCD charges from int to FixedPoint2


## Objects and Categories:

### Piping (1/2 Charge):
- Fourway
- Straight
- TJunction
- Bend

### Pumps & Valves (1 Charge):

- Pressure Pump
- Volumetric Pump
- Signal Valve
- Manual Valve 
- Pneumatic Valve 


### Atmospheric Utility (1 Charge):
- Gas Mixer
- Gas Filter
- Canister Port
- Radiator


### Vents:
- Outlet Injector
- Scrubber Vent
- Passive Vent
- Air Vent
- Dual Port

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->


https://github.com/user-attachments/assets/6a4b431e-5722-4d4e-9e00-00afcd5af926

The sprites fitnow ^

![image](https://github.com/user-attachments/assets/90823f50-6a0e-4ec3-8d6d-1c5b4bf84bd8)
![image](https://i.imgur.com/WKPAg00.png)

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->

:cl:
- add: Rapid piping device.

gusxyz committed:
initial rpd

gusxyz committed:
added and reorganized

gusxyz committed:
added RCDDeconstructable componenet to GasPipeBase

ArtisticRoomba commented:
oh my god it's beautiful

Ilya246 commented:
should maybe ideally take steel and not RCD charges, i'd also suggest for it to have 2 charges per unit steel

ArtisticRoomba commented:
> should maybe ideally take steel and not RCD charges, i'd also suggest for it to have 2 charges per unit steel
hard agree, please do this
make sure the steel charges balance out with the cost of objects that take 2 steel

chromiumboy commented:
Could you please list all the the structures and features you plan on implementing for the RPD in the description? It will make giving feedback easier

TurboTrackerss14 commented:
Oh swear on the lords above if this doesn't have checks for valid piping to prevent pipe stacking I am going to morb. 

Additionally keep it as RCD charges over charges per steel ™️ 
it works for the RCD it should work here.

Ilya246 commented:
> Oh swear on the lords above if this doesn't have checks for valid piping to prevent pipe stacking I am going to morb.

pipes have the checks so not even the entity spawn menu can stack pipes

TurboTrackerss14 commented on Content.Client/RCD/RCDMenu.xaml:
```diff
@@ -40,8 +52,20 @@
+    <!-- Atmospheric Utility -->
+    <ui:RadialContainer Name="AtmosphericUtility"  VerticalExpand="True" HorizontalExpand="True" Radius="64"/>
+
+    <!-- Atmospheric Utility -->
```
Copy paste typo? 
Should likely be 
<!-- Vents -->

TurboTrackerss14 commented on Resources/Prototypes/RPD/rpd.yml:
```diff
@@ -0,0 +1,189 @@
+  rotation: User
+  fx: EffectRCDConstruct0
+
+- type: rcd
```
Needs   rules:
  - MustBuildOnSubfloor same as the other pipes

keronshb commented:
With the RPD we also wanted to introduce layered piping. So this will pave a nice pathway to that.

gusxyz commented:
> With the RPD we also wanted to introduce layered piping. So this will pave a nice pathway to that.
> 
> Things to keep in mind:
> 
> * Pipes shouldn't be able to anchor on top of other pipes. You should be able to do like, a cross? So a horizontal and a vertical pipe on the same tile, but not being able to stack pipes.
> * It's fine to include atmos devices (pumps, valves, etc)
> * One device per tile
> * If colored piping is supported, a way to color pipes
> * A toggle to spawn a pipe anchored or not (default anchored)
> 
> I feel like most of these are already handled by the systems we have in place already but just things to keep in mind when testing.

Im 95% sure the RCD and piping/pipenet devices covers everything besides the painting and toggling of anchoring.

gusxyz commented:
> Could you please list all the the structures and features you plan on implementing for the RPD in the description? It will make giving feedback easier

Ill add it to the pr

gusxyz commented:
I updated the pull request with what I've implemented and plan to implement.

Please remind me if I forgot anything or could reorganize the categories better.

gusxyz committed:
fixed typo and a lack of MustbuildONSubfloor components.

slarticodefast commented:
> If colored piping is supported, a way to color pipes

It should be one color for each pipe layer  instead of using the spray painter imo.

qwerltaz commented:
Add pneumatic valve and dual-port vent to available objects.

Originally the rpd supports placing disposal pipes and junctions too. We could consider adding them here as well.

Given the number of items and later more options with layers, especially if disposal pipes are added to the rpd, you could consider a window with a texture button grid for choosing the entity and a column with number buttons for choosing pipe layer instead of radial menu, similar to crayon decal choice or decal placement sandbox menu.

cohanna commented:
> Additionally keep it as RCD charges over charges per steel ™️ it works for the RCD it should work here.

utter waste of compressed matter, the utility of the item is killed if it uses something as limited as compressed matter. If the cost per pipe is even a single charge the rpd won’t even have enough charges to setup waste on fland. Why would any atmos use their compressed matter on the RPD instead of the RCD and construct the pipes by hand. It’s like how almost no engines rely on the RCD to place cable. Despite being cheap it’s a waste when it’s easy enough to do by hand.

keronshb commented:
To be fair, the RPD had infinite charges over on SS13. Issue of having infinite ammo here is simply the amount of entities someone can spam.

cohanna commented:
> To be fair, the RPD had infinite charges over on SS13. Issue of having infinite ammo here is simply the amount of entities someone can spam.

I'm fine with it not being infinite, but compressed matter is way too expensive and limited to be used as ammo for the RPD (even if it can be bought from cargo). Slotting steel into it is a much better solution imo.

ArtisticRoomba commented:
Only slightly related but IMO engineering should be able to produce compressed matter using a lot of steel and glass. Maybe in a "bluespace press" machine or something. If that was already implemented I could understand the compressed matter as ammo argument but for now I think steel as ammo is a good compromise.

chromiumboy commented:
Just to clarify a few things brought up here:
- The RCD was never intended to be efficient. It is supposed to be quick, convenient, and provides you with all the things you need to restore basic station infrastructure and function (sans atmospherics)
- If the RPD is to use the RCD framework, compressed matter makes sense as the ammo type
- The number of charges on the RPD can be adjusted to ensure that atmos can be set up
- The RCD system does not support arbitrary materials as a resource, it would need to be reworked to accommodate them
- SS13 also had the pipe dispenser, which was a stationary structure; this could be implemented later as the more efficient but less convenient alternative to the RPD

cohanna commented:
> - The RCD was never intended to be efficient. It is supposed to be quick, convenient, and provides you with all the things you need to restore basic station infrastructure and function (sans atmospherics)

Isn't the RPD supposed to be efficient though? The entire point of it is so you don't have to use the construction menu to setup pipes. 9 times out of 10 pipe repairs are typically only 3 or 4 damaged pipes (syndiebombs aside) which is trivial to repair even with the current system. Now if this is talking about the compressed matter being much better utilized for the RCD I direct you to its unique deconstruction abilities.

> - If the RPD is to use the RCD framework, compressed matter makes sense as the ammo type
> - The RCD system does not support arbitrary materials as a resource, it would need to be reworked to accommodate them

I mean that sucks, yeah, but compressed matter doesn't make sense in its current form/cost for the RPD

chromiumboy commented:
> Isn't the RPD supposed to be efficient though? The entire point of it is so you don't have to use the construction menu to setup pipes.

I meant that it isn't meant to be material efficient. Time efficient, yes, absolutely

cohanna commented:
> I meant that it isn't meant to be material efficient. Time efficient, yes, absolutely

Then it can use an inefficient amount of steel, but compressed matter still doesn't make sense unless it uses a fraction of a charge per pipe or something.

gusxyz commented:
I think the best solution for compressed matter vs steel issue is to just make a way to create compressed matter at a lathe (possibly through research)

In the yml compressed matter it says its composed of a sheet of plastic and steel..

the only issue I could see arising is a potential balance issues.

Ilya246 commented:
the RPD is big for atmos accessibility and i think it's fine if it's just straight up better than manual pipe construction, having it be both fast and efficient would allow atmos construction to no longer feel clunky without also being inefficient

gusxyz commented:
I was able to setup fland's distro with 10 charges left with the current compressed matter requirements.

Considering the fact at round start your locker contains an extra compressed matter means you have more than enough to setup the waste.

Keep in mind cargo can order compressed matter at the moment.

gusxyz committed:
reorganized categories, added pneumatic valve, signal valve and dual airport

gusxyz commented:
Another option we have is to increase the RPD's compressed matter storage and starting amount to double of the RCD since it seems visually larger.

Ilya246 commented:
i think having it use compressed matter is only acceptable if it can build multiple pipes per matter unit

cohanna commented:
> I was able to setup fland's distro with 10 charges left with the current compressed matter requirements.
> 
> 
> 
> Considering the fact at round start your locker contains an extra compressed matter means you have more than enough to setup the waste.
> 
> 
> 
> Keep in mind cargo can order compressed matter at the moment.


Again, who in their right mind would use that extra charge for the RPD over the RCD. Furthermore, if you are utilizing the vast majority of your charges just to do distro and waste (arguably the most simple, quick thing to do) then the tool doesn't meet the goal (correct me if I'm wrong here) of "replacing the construction menu" and wouldn't be used for actual atmos work until cargo forks over 2.5k for some matter. (Aka not worth it when steel is easy enough to get by deconstructing some lockers)
Having it rely on compressed matter doesn't allow it to be used as the preferable alternate to the construction menu which means it's mostly useless.

cohanna commented:
Not to mention that placing incorrect pipes would be pretty punishing since it would cost compressed matter charges to place, deconstruct, and place again. Which, again, makes it ill suited to replace the construction menu for atmos techs. 
I mean why wouldn't you immediately switch back to the construction menu as soon as you make a mistake? At least you can get your materials back if you avoid the RPD.

If instead it uses steal as charges deconstructing a misplaced pipe by hand will allow the RPD to continue to be utilized since you can stick the steel right back in.

gusxyz commented:
> i think having it use compressed matter is only acceptable if it can build multiple pipes per matter unit

I'm planning to double the RPD's charges, it will be basically the same thing.

gusxyz commented:
the other option making the rpd rechargeable like the bluespace rcd

gusxyz committed:
converted rcd charges values to float, made pipes a half a charge and, made rpd capacity 45.

gusxyz commented:
also no c# label is invalid now.

Ilya246 commented:
i vote for just going the ss13 route and making pipes from RPD free (self charging) but also have pipes deconstruct into nothing

iaada commented:
I like it using compressed matter. Cargo being bad and not being able to buy more matter isn't really an issue with the RPD, it's an issue with cargo. I think so long as mappers replace the 30 steel in the atmos work room with an extra stack of matter, there shouldn't be any complaints on wasting charges for round start.

TurboTrackerss14 commented:
> i vote for just going the ss13 route and making pipes from RPD free (self charging) but also have pipes deconstruct into nothing
Highly unlikely to be accepted 
Source: infinite chem got axed and at the same time "energy based" chem production was given a no (to my recollction). 

Pipes are much more physical than a random jug of chems. They should cost real metal and give said metal back in return for deconsturction.
The RPD should use matter additionally for parity between RCD and RPD ™️

marc-pelletier commented:
Just make it cost steel. Steel is easy enough to find as is, and if it comes loaded with a stack, atmos is already ahead. Won't have to spend time figuring out how to balance material costs, or have to mess with the way pipes work to make them return nothing. Everything keeps it's current balance and atmos is less tedious to set up.

gusxyz committed:
FixedPoint2 instead of float.

YotaXP commented:
For the compressed matter vs steel debate, let me point out the following.

- The RCD builds things using a mix of steel, glass, and even plastic for door electronics. Additionally, it has a special function for deleting a variety of things. It makes sense for it to require a special ammo type.

- The RPD, on the other hand, exclusively **builds** things that are made out of steel. It makes sense for it to consume steel directly. Furthermore, mistakes would be extra punishing if you couldn't just melt down the misplaced piping and load the steel back into the device.

On that note, the deconstruct mode should probably be either removed, or limited to piping and return materials used.

lzk228 commented on Resources/Prototypes/Entities/Objects/Tools/tools.yml:
```diff
@@ -337,6 +337,56 @@
 
+- type: entity
+  id: RPD
+  parent: [ BaseItem, BaseEngineeringContraband ]
```
use rcd as a parent, then you can decrease the number of lines

gusxyz replied:
will do

lzk228 commented on Resources/Prototypes/Entities/Objects/Tools/tools.yml:
```diff
@@ -337,6 +337,56 @@
+  - type: LimitedCharges
+    maxCharges: 45
+    charges: 45
+  - type: UseDelay
```
then you can remove this

lzk228 commented on Resources/Prototypes/Entities/Objects/Tools/tools.yml:
```diff
@@ -337,6 +337,56 @@
+  - type: UseDelay
+  - type: Sprite
+    sprite: Objects/Tools/rpd.rsi
+    state: icon
```
the state field

lzk228 commented on Resources/Prototypes/Entities/Objects/Tools/tools.yml:
```diff
@@ -337,6 +337,56 @@
+    sprite: Objects/Tools/rpd.rsi
+    state: icon
+  - type: Item
+    size: Normal
```
this component and all below this also will inherit from rcd

lzk228 commented on Resources/Prototypes/Entities/Structures/Piping/Atmospherics/pipes.yml:
```diff
@@ -44,6 +48,7 @@
         offset: 0
       - !type:DoActsBehavior
         acts: ["Breakage"]
+
```
why?

gusxyz replied:
what are you asking ?

gusxyz replied:
if its the linebreak that was def an accident

lzk228 commented on Resources/Textures/Interface/Radial/RPD/Straight.png:
why don't you use the current sprites?

gusxyz replied:
look at the rcds same sprite they are flat 1 directional 16x16 images and i have no intent to touch radial ui, its out of the scope of this pr

gusxyz committed:
Fixed linebreak typo and made RPD a child of RCD.

gusxyz committed:
fixed filename typo

lzk228 commented on Resources/Prototypes/Entities/Objects/Tools/tools.yml:
```diff
@@ -339,7 +339,7 @@
 - type: entity
   id: RPD
-  parent: [ BaseItem, BaseEngineeringContraband ]
+  parent: [ BaseItem, BaseEngineeringContraband, RCD]
```
bruh
```suggestion
  parent: RCD
```

lzk228 commented on Resources/Prototypes/Entities/Objects/Tools/tools.yml:
```diff
@@ -367,25 +367,10 @@
-  - type: UseDelay
   - type: Sprite
     sprite: Objects/Tools/rpd.rsi
     state: icon
```
this line also inherits

lzk228 reviewed (changes requested):
use snake_case for naming sprites

gusxyz committed:
snake_case and fixed the parenting.

gusxyz committed:
forgot RCDMenu.xaml and removed the not needed icon state line.

YotaXP commented:
For parity with construction ghosts, the RPD should be able to build on tiled floors.

https://github.com/user-attachments/assets/840c7086-dde3-4587-b526-6af2a1d234b4

qwerltaz commented:
>  How do you flip the build for ternary devices? Ctrl+Middle Click doesn't seem to be doing it. Instead it throws me into construction ghost mode if the last buildable I had selected was flippable.

Looks like the flip message is intercepted by `Content.Client.Construction.ConstructionSystem.HandleFlip` because it directly listens to the keybind. Preferably we can have it fixed here so filters, mixers and similar can be flipped when using rpd.

southbridge-fur commented:
> For parity with construction ghosts, the RPD should be able to build on tiled floors.
> 2024-10-08.21-24-19.mp4
> 
> Edit: How do you flip the build for ternary devices? Ctrl+Middle Click doesn't seem to be doing it. Instead it throws me into construction ghost mode if the last buildable I had selected was flippable. Also, half pipes are missing. (Though I don't really know why those exist in the first place.)

Annoyingly, even in regular build mode you can't flip ternary devices. You have to construct the device, un-anchor it, flip it, then re-anchor it.

gusxyz commented:
> > For parity with construction ghosts, the RPD should be able to build on tiled floors.
> 
> > 2024-10-08.21-24-19.mp4
> 
> > 
> 
> > Edit: How do you flip the build for ternary devices? Ctrl+Middle Click doesn't seem to be doing it. Instead it throws me into construction ghost mode if the last buildable I had selected was flippable. Also, half pipes are missing. (Though I don't really know why those exist in the first place.)
> 
> 
> 
> Annoyingly, even in regular build mode you can't flip ternary devices. You have to construct the device, un-anchor it, flip it, then re-anchor it.

thanks to @chromiumboy being the literal goat flipping will be getting implemented as soon as i can.

gusxyz commented:
> For parity with construction ghosts, the RPD should be able to build on tiled floors.
> 
> 
> 
> https://github.com/user-attachments/assets/840c7086-dde3-4587-b526-6af2a1d234b4
> 
> 
> 
> Edit:
> 
> How do you flip the build for ternary devices? Ctrl+Middle Click doesn't seem to be doing it. Instead it throws me into construction ghost mode if the last buildable I had selected was flippable.
> 
> Also, half pipes are missing. (Though I don't really know why those exist in the first place.)
> 
> 
Ill add the half pipes cause im sure someone has use for them..

gusxyz committed:
added flipping thanks chromium <3

gusxyz commented:
https://github.com/user-attachments/assets/7e08ead8-fa9c-4ff5-bf9e-f90dfe8af582

Centronias commented on Resources/Prototypes/RPD/rpd.yml:
```diff
@@ -0,0 +1,259 @@
+  category: Piping
+  sprite: /Textures/Interface/Radial/RPD/straight.png
+  mode: ConstructObject
+  prototype: GasPipeStraight
+
+  cost: 0.5
```
```suggestion
  prototype: GasPipeStraight
  cost: 0.5
```

Centronias reviewed (commented):
Might wanna go through and doublecheck all of the whitespace changes to make sure they're intentional.

qwerltaz commented:
Deconstructing pipes with the rpd takes a long time. It should be instant just like when placing them.

The rpd can deconstruct plating if you press 1 pixel outside of a pipe which can lead to frequent spacing by mistake. It probably shouldn't be able do decon plating, or maybe anything else like doors, walls and windows.

Feels like rpd should have higher range than the rcd. Now it's 1 so you have to stand right beside where you want to place, which is annoying for placing pipes.

chromiumboy commented:
If the RPD has the ability to deconstruct, it should only be able to deconstruct pipes and associated atmos devices

chromiumboy commented:
Test fails are real

gusxyz committed:
fixed whitespace issues

gusxyz committed:
fixed more whitespace issues

gusxyz commented:
> If the RPD has the ability to deconstruct, it should only be able to deconstruct pipes and associated atmos devices

working on this

gusxyz committed:
Added RPD whitelist to pipes. RPD now can only destroy piping and related devices.

gusxyz commented:
> Test fails are real

why wont it recognize my .png files that are there

gusxyz commented:
> > Test fails are real
> 
> why wont it recognize my .png files that are there

it wont update the capitalization..

gusxyz committed:
fixed ignore case and removed a debug message

jamessimo commented:
Is there a reason to duplicate the textures into the interface folder? I assume RCD does this too but I think it would be better to just refence the texture that will be placed like the construction menu uses. Will make skinning objects annoying I think + future QOL could be to also set colors for the pipes.

gusxyz commented:
> Is there a reason to duplicate the textures into the interface folder? I assume RCD does this too but I think it would be better to just refence the texture that will be placed like the construction menu uses. Will make skinning objects annoying I think + future QOL could be to also set colors for the pipes. 

afaik the interface sprites dont work with 4 directional sprites unless a maintainer can correct me.

marc-pelletier commented:
After playing around with the RPD a bit, condensed matter probably works fine mechanically. It's also just more consistent with in game lore.

One thing that I noticed is that you can't overlap pipes across one another. For instance, two straight pipes oriented in different directions, north-south and east-west, won't overlap. Same with two bend pipes oriented for instance northwest and southeast.

gusxyz commented:
> -

Thanks for this testing marc!

gusxyz commented:
> After playing around with the RPD a bit, condensed matter probably works fine mechanically. It's also just more consistent with in game lore.
> 
> One thing that I noticed is that you can't overlap pipes across one another. For instance, two straight pipes oriented in different directions, north-south and east-west, won't overlap. Same with two bend pipes oriented for instance northwest and southeast.
> 
> EDIT: Another issue I discovered is that all non-pipes are stackable. Connector ports, filters, scrubbers, etc.

Thank you for testing this, I'll look into it when I have some free time.

gusxyz commented:
triaged yay

qwerltaz commented on Resources/Prototypes/RPD/rpd.yml:
```diff
@@ -0,0 +1,258 @@
+  rotation: User
+  fx: EffectRCDConstruct0
+
+- type: rcd
+  id: VolumetricPump
+  category: PumpsValves
+  sprite: /Textures/Interface/Radial/RPD/pump_volume.png
+  mode: ConstructObject
+  prototype: GasPressurePump
```
This field for the volumetric pump spawns a normal pressure pump instead due to wrong prototype.

qwerltaz replied:
This field for the volumetric pump spawns a normal pressure pump instead due to wrong prototype.

Pumkin69 commented:
RAPID PIPING DEVICE REAL!!!
Genuinely great pr thank you

gusxyz committed:
you can build under tiles now and fixed volume pipe, thanks marc

YotaXP commented:
Bent radiators will need to be added to the RPD: #34251

MilonPL commented on Content.Client/RCD/RCDConstructionGhostSystem.cs:
```diff
@@ -1,11 +1,15 @@
 using Robust.Shared.Enums;
+using Robust.Shared.Input;
+using Robust.Shared.Input.Binding;
+
```
Remove the extra newline.

MilonPL commented on Content.Client/RCD/RCDConstructionGhostSystem.cs:
```diff
@@ -17,6 +21,52 @@ public sealed class RCDConstructionGhostSystem : EntitySystem
 
     private string _placementMode = typeof(AlignRCDConstruction).Name;
     private Direction _placementDirection = default;
+    private bool _useMirrorPrototype = false;
```
`default` and `false` are already being set for these, you don't need the extra initialization.

MilonPL commented on Content.Client/RCD/RCDConstructionGhostSystem.cs:
```diff
@@ -17,6 +21,52 @@ public sealed class RCDConstructionGhostSystem : EntitySystem
     private string _placementMode = typeof(AlignRCDConstruction).Name;
     private Direction _placementDirection = default;
+    private bool _useMirrorPrototype = false;
+    public event EventHandler? FlipConstructionPrototype;
```
What is this for? This is never invoked.

MilonPL commented on Content.Client/RCD/RCDConstructionGhostSystem.cs:
```diff
@@ -55,20 +105,28 @@ public override void Update(float frameTime)
-        // If the placer has not changed, exit
+        // If the placer has not changed build it.
         _rcdSystem.UpdateCachedPrototype(heldEntity.Value, rcd);
+        var useProto = (_useMirrorPrototype && !string.IsNullOrEmpty(rcd.CachedPrototype.MirrorPrototype)) ? rcd.CachedPrototype.MirrorPrototype : rcd.CachedPrototype.Prototype;
```
Really long line, could be split over multiple.

MilonPL commented on Content.Client/RCD/RCDConstructionGhostSystem.cs:
```diff
@@ -55,20 +105,28 @@ public override void Update(float frameTime)
+
+        if (heldEntity != placerEntity || useProto != placerProto)
+        {
+            CreatePlacer(heldEntity.Value, rcd, useProto);
```
Could use pattern matching, even somewhere above, where we're first declaring `heldEntity`.

MilonPL commented on Content.Client/RCD/RCDConstructionGhostSystem.cs:
```diff
@@ -55,20 +105,28 @@ public override void Update(float frameTime)
 
+    }
+
+    private void CreatePlacer(EntityUid uid, RCDComponent component, string? prototype)
```
`Entity<T>` pattern here

MilonPL commented on Content.Shared/RCD/Components/RCDComponent.cs:
```diff
@@ -42,6 +42,18 @@ public sealed partial class RCDComponent : Component
+    /// <summary>
+    /// Indicates whether this is an RCD or an RPD
+    /// </summary>
+    [DataField, AutoNetworkedField]
```
Does this need to be networked?

metalgearsloth replied:
This comment is entirely undescriptive on what it even does and is a code smell that it's a bool toggle to do hidden behaviors.

This goes against data-driven design and instead RCD itself should just be able to swap between the 2 without requiring hidden toggles.

MilonPL commented on Content.Shared/RCD/Components/RCDComponent.cs:
```diff
@@ -42,6 +42,18 @@ public sealed partial class RCDComponent : Component
+    /// Indicates whether this is an RCD or an RPD
+    /// </summary>
+    [DataField, AutoNetworkedField]
+    public bool IsRpd { get; set; } = false;
```
This doesn't need a get/set, `false` is default.

MilonPL commented on Content.Shared/RCD/Components/RCDComponent.cs:
```diff
@@ -42,6 +42,18 @@ public sealed partial class RCDComponent : Component
+    /// Indicates if a mirrored version of the construction prototype should be used (if available)
+    /// </summary>
+    [AutoNetworkedField, ViewVariables(VVAccess.ReadOnly)]
+    public bool UseMirrorPrototype = false;
```
Shouldn't this be a DataField?

gusxyz replied:
Negative.
Datafield is for yml defined variables.
all good.

metalgearsloth replied:
You're networking it by definition it should be a datafield so the state saves properly.

MilonPL commented on Content.Shared/RCD/Components/RCDDeconstructibleComponent.cs:
```diff
@@ -31,4 +32,11 @@ public sealed partial class RCDDeconstructableComponent : Component
+    /// Toggles whether this entity is deconstructable by the RPD or not
+    /// </summary>
+    [DataField("rpd"), ViewVariables(VVAccess.ReadWrite)]
+    public bool RpdDeconstructable = false;
```
Doesn't need the `ViewVariables` attribute, false is the default for bools.

MilonPL commented on Content.Shared/RCD/RCDPrototype.cs:
```diff
@@ -44,11 +45,17 @@ public sealed partial class RCDPrototype : IPrototype
+    /// If the entity can be flipped, this prototype is available as an alternate (mode dependent)
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadOnly)]
+    public string? MirrorPrototype { get; private set; } = string.Empty;
```
I know the field above also uses a string, but both of these should be turned into `EntProtoId`.

MilonPL commented on Content.Shared/RCD/Systems/RCDSystem.cs:
```diff
@@ -25,6 +25,7 @@
 using Robust.Shared.Timing;
 using System.Diagnostics.CodeAnalysis;
 using System.Linq;
+using Content.Shared.FixedPoint;
```
Sort this with the other usings.

MilonPL commented on Content.Shared/RCD/Systems/RCDSystem.cs:
```diff
@@ -65,6 +66,8 @@ public override void Initialize()
         SubscribeLocalEvent<RCDComponent, RCDSystemMessage>(OnRCDSystemMessage);
         SubscribeNetworkEvent<RCDConstructionGhostRotationEvent>(OnRCDconstructionGhostRotationEvent);
+        SubscribeNetworkEvent<RCDConstructionGhostFlipEvent>(OnRCDConstructionGhostFlipEvent);
+
```
Remove the extra newline.

metalgearsloth replied:
This should be using the predicted version of subscription.

MilonPL commented on Content.Shared/RCD/Systems/RCDSystem.cs:
```diff
@@ -85,6 +88,8 @@ private void OnMapInit(EntityUid uid, RCDComponent component, MapInitEvent args)
         QueueDel(uid);
     }
 
+
```
Remove the extra newlines.

MilonPL commented on Content.Shared/RCD/Systems/RCDSystem.cs:
```diff
@@ -482,16 +516,25 @@ private bool IsDeconstructionStillValid(EntityUid uid, RCDComponent component, M
         else
         {
+            // The object is not in the RPD whitelist
+            if (!TryComp<RCDDeconstructableComponent>(target, out var deconstructible) || !deconstructible.RpdDeconstructable && component.IsRpd)
```
Really long line.

MilonPL commented on Content.Shared/RCD/Systems/RCDSystem.cs:
```diff
@@ -589,8 +637,12 @@ private bool DoesCustomBoundsIntersectWithFixture(PolygonShape boundingPolygon,
     {
-        if (component.ProtoId.Id != component.CachedPrototype?.Prototype)
+        if (component.ProtoId.Id != component.CachedPrototype?.Prototype ||
+            (component.CachedPrototype?.MirrorPrototype != null &&
```
CachedPrototype is never null, this doesn't need conditional access.

MilonPL commented on Content.Shared/RCD/Systems/RCDSystem.cs:
```diff
@@ -589,8 +637,12 @@ private bool DoesCustomBoundsIntersectWithFixture(PolygonShape boundingPolygon,
-        if (component.ProtoId.Id != component.CachedPrototype?.Prototype)
+        if (component.ProtoId.Id != component.CachedPrototype?.Prototype ||
+            (component.CachedPrototype?.MirrorPrototype != null &&
+             component.ProtoId.Id!= component.CachedPrototype?.MirrorPrototype))
```
What's going on here? Why are you comparing a bool to a string?

MilonPL commented on Resources/Prototypes/Entities/Objects/Tools/tools.yml:
```diff
@@ -337,6 +337,41 @@
     inHandsOnly: true
     key: enum.RcdUiKey.Key
 
+- type: entity
```
Yaml conventions, the order should be `parent` > `id` > `name` > `description` > `components`

MilonPL reviewed (commented):
Like someone else said, also needs bent radiators added.

gusxyz commented:
@MilonPL THANK YOU FOR TAKING TIME TO REVIEW THE CODE <3

ilovehans10 commented:
This will crash the client without the new radial menu syntax in RCDMenu.xaml. I've fixed this in a [PR downstream](https://github.com/ss14-harmony/ss14-harmony/pull/370), and if you want you can copy the change I did there. This is partially selfish, as it would have the added benefit of making it easier for us to merge with upstream when this is merged.

gusxyz commented:
> This will crash the client without the new radial menu syntax in RCDMenu.xaml. I've fixed this in a [PR downstream](https://github.com/ss14-harmony/ss14-harmony/pull/370), and if you want you can copy the change I did there. This is partially selfish, as it would have the added benefit of making it easier for us to merge with upstream when this is merged.

thank you

keronshb commented:
@gusxyz Can you fix the conflicts and add the new radial menu?

gusxyz commented:
😄 yeah soon

metalgearsloth commented on Content.Client/RCD/RCDConstructionGhostSystem.cs:
```diff
@@ -17,6 +21,52 @@ public sealed class RCDConstructionGhostSystem : EntitySystem
+
+            _useMirrorPrototype = !rcd.UseMirrorPrototype;
+
+            var useProto = _useMirrorPrototype ? rcd.CachedPrototype.MirrorPrototype : rcd.CachedPrototype.Prototype;
+            CreatePlacer(placerEntity.Value, rcd, useProto);
+
+            // tell the server
+
+            RaiseNetworkEvent(new RCDConstructionGhostFlipEvent(GetNetEntity(placerEntity.Value), _useMirrorPrototype));
```
This should be predicted.

metalgearsloth commented on Content.Client/RCD/RCDConstructionGhostSystem.cs:
```diff
@@ -17,6 +21,52 @@ public sealed class RCDConstructionGhostSystem : EntitySystem
+            if(!TryComp<RCDComponent>(placerEntity, out var rcd) ||
+                string.IsNullOrEmpty(rcd.CachedPrototype.MirrorPrototype))
+                return false;
+
+            _useMirrorPrototype = !rcd.UseMirrorPrototype;
```
You shouldn't be storing stateful data on systems.

ilovehans10 commented:
We have been enjoying the RPD on Harmony for a while now, and I just wanted to make sure that it was on your radar that when we merged upstream it looks like #34639 made it so the RPD menu lost all actions but deconstruct. This was fixed by a wonderful member of our community in [this PR](https://github.com/ss14-harmony/ss14-harmony/pull/519).

gusxyz commented:
I'll get back to this soon. I've been beating my head over making RPD's supercode comply with rcd refactor.

gusxyz commented:
> Probably beating a dead horse here
> 
> If I had a choice between using RCD charges to make pipes vs just building them the standard way with steel it becomes very obvious that building them with steel is far more cost effective.
> 
> RCD gets away with this material inefficiency because it's typically more for emergencies or the convenience of not needing complex construction/niche parts such as doors or electrical wires.
> 
> The RPD is almost entirely for convenience, so I think it's fair to either make it more efficient with its charges or allow it to take steel instead of RCD charges(all its constructions only use steel anyways)
> 
> As it stands if I were CE I'd be legitimately upset to see someone wasting RCD charges on pipes rather than using them for emergency situations.  Won't be too bad once department economy comes, CE could just buy more, but it goes to show it's significantly less useful than an RCD despite using the same resources.
> 
> Edit: I take back saying it was less effective, not exactly what I meant to convey, I don't want it to seem like I don't love this and welcome it into the game.  

Your beating a dead horse dude, downstreams have had RPDs with 0 complaints for many months.

Pumkin69 commented:
We should really just have a way to turn materials into compressed matter via a machine from science or something at some point as well.

keronshb commented:
Would you also be able to add pipe layering to this as well?

ilovehans10 commented:
I dream that some day this will exist on upstream.

Zarq0 commented:
sadge
